Issue 1402 - implement divmod, even, odd

winitzki opened the pull request:
This PR implements `Natural/even` and `Natural/odd` via `Natural/divMod`. The algorithm for `divMod` is efficient even for very large `Natural` numbers, unlike algorithms based on simple subtraction.

`Function/Validate` is a utility for creating generic validation functions, such as, to guarantee that `divMod y {=} x` is never called to divide by `y = 0`. 

See the discussion here https://github.com/dhall-lang/dhall-lang/issues/1213 about the details of how `divMod` statically avoids division by zero.

winitzki committed:
implement Natural/showHex in Prelude

winitzki committed:
wip

winitzki committed:
implement even, odd

winitzki committed:
bump ci

winitzki committed:
some more comments

winitzki commented:
Not sure why tests are failing here. Logs say "timed out" and no error messages.

winitzki committed:
bump ci

winitzki committed:
Merge branch 'master' into feature/issue-1402-implement-divmod-even-odd

nikita-volkov commented:
Nice work!

Dependent types however seems like a philosophical change for the base library, since no other operation in it relies on dependent types. It also immediately raises the bar for the possible users automatically decreasing adoption. I suspect this is the reason why this hasn't been merged so far.

I suggest reimplementing this as `divMod : Natural -> Natural -> Optional Result`.

winitzki committed:
Merge branch 'master' into feature/issue-1402-implement-divmod-even-odd

winitzki committed:
Enhance documentation for Validate functionality

Added documentation for the Validate function, explaining its purpose and usage with examples.

nikita-volkov reviewed (changes requested):
Let's discuss introducing a divMod primitive
